scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py

```diff
@@ -0,0 +1,485 @@
+        print("|{0:^7}|{1:^7}|{2:^7}|{3:^10}|{4:^10}|{5:^10}|{6:^10}|"
+              .format("niter", "f evals", "CG iter", "tr radius", "penalty",
+                      "opt", "c viol"))
+        s = "-"*6 + ":"
+        s2 = ":" + "-"*8 + ":"
+        print("|{0:^7}|{1:^7}|{2:^7}|{3:^10}|{4:^10}|{5:^10}|{6:^10}|"
+              .format(s, s, s, s2, s2, s2, s2))
+
+    @staticmethod
+    def print_problem_iter(niter, nfev, cg_niter, tr_radius,
+                           penalty, opt, c_viol):
+        print("|{0:>7}|{1:>7}|{2:>7}| {3:^1.2e} | {4:^1.2e} |"
+              " {5:^1.2e} | {6:^1.2e} |"
+              .format(niter, nfev, cg_niter, tr_radius, penalty,
+                      opt, c_viol))
+
+    @staticmethod
+    def print_footer():
+        print("")
+        print((7*3 + 10*4 + 8)*"-")
+        print("")
+
+
+class ip_printer:
+    @staticmethod
+    def print_header():
+        print("|{0:^7}|{1:^7}|{2:^7}|{3:^13}|{4:^10}|{5:^10}|{6:^10}|{7:^10}|"
+              .format("niter", "f evals", "CG iter", "barrier param",
+                      "tr radius", "penalty", "opt", "c viol"))
+        s = "-"*6 + ":"
+        s2 = ":" + "-"*11 + ":"
+        s3 = ":" + "-"*8 + ":"
+        print("|{0:^7}|{1:^7}|{2:^7}|{3:^13}|{4:^10}|{5:^10}|{6:^10}|{7:^10}|"
+              .format(s, s, s, s2, s3, s3, s3, s3))
+
+    @staticmethod
+    def print_problem_iter(niter, nfev, cg_niter, barrier_parameter, tr_radius,
+                           penalty, opt, c_viol):
+        print("|{0:>7}|{1:>7}|{2:>7}|   {3:^1.2e}  | "
+              "{4:^1.2e} | {5:^1.2e} | {6:^1.2e} | {7:^1.2e} |"
+              .format(niter, nfev, cg_niter, barrier_parameter,
+                      tr_radius, penalty, opt, c_viol))
+
+    @staticmethod
+    def print_footer():
+        print("")
+        print((7*3 + 13 + 10*4 + 9)*"-")
+        print("")
+
+
+def _minimize_trustregion_constr(fun, x0, args, grad,
+                                 hess, hessp, constraints,
+                                 xtol=1e-8, gtol=1e-8,
+                                 barrier_tol=1e-8,
+                                 sparse_jacobian=None,
+                                 callback=None, maxiter=1000,
+                                 verbose=0, finite_diff_options=None,
+                                 initial_penalty=1.0, initial_trust_radius=1.0,
+                                 initial_barrier_parameter=0.1,
+                                 initial_tolerance=0.1,
+                                 return_all=False, factorization_method=None,
+                                 disp=False):
+    """Minimize scalar function subject to constraints.
+
+    Options
+    -------
+    gtol : float, optional
+        Tolerance for termination by the norm of the Lagrangian gradient.
+        The algorithm will terminate when both the infinity norm (i.e. max
+        abs value) of the Lagrangian gradient and the constraint violation
+        are smaller than ``gtol``. Default is 1e-8.
```

Comment:
Why `gtol` is used for both Lagrangian gradient and constraint violation?

Reply:
I have been using this termination condition so far and it have worked well enough. The original paper also formulates the terminal condition in this way.

Do you think I should separate both conditions? In truth I do not have a strong opinion about this, so if you think one of the alternatives is better, just say and I will modify it.

Reply:
To be honest I don't know. What concerns me here is that thinking in terms of physical quantities --- the gradient of the Lagrangian and constraints themselves generally have different units.

Say, the independent variables has dimension "meters" (some distances), objective function has units "grams" (optimizing some mass), and the constraints are dimensional constraints having units "meters" as well. So the gradient of the Lagrangian has units "grams / meters", whereas the constraints has units "meters". From this perspective it doesn't look right.

But, on the other hand, if we consider this as just a mathematical problem (which is probably wiser thing to do) and also this is suggested in the paper and works well, I suggest to keep it that way.

Reply:
Ok your argument make sense to me.

I will keep as it is for now but keep your argument in mind and revisit this latter after the rest of the code have been refactored.